### PR TITLE
script: Linter to check commit message formatting

### PR DESCRIPTION
**Pull request by am1r021**
Write linter to check that commit messages have a new line before the body or no body at all. fixes issue #19091.

### CODE REVIEW

File: test/lint/lint-git-commit-check.sh [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    n_line=0
+    while IFS= read -r line || [[ -n "$line" ]]; do
+        n_line=$((n_line+1))
+        length=${#line}
+        if [ $n_line -eq 2 ] && [ $length -ne 0 ]; then
+            echo "Your subject line follows a non-empty line. Subject lines should always be followed by a blank line."
+            exit 1
+        fi
+    done < <(echo "$msg")
+done < <(git log origin..HEAD --format=%s%b+)
```
Comment (adamjonas) [inline test/lint/lint-git-commit-check.sh:25]: `$ git log origin..HEAD --format=%s%b+` isn't working for me while testing locally. I get:
```
fatal: ambiguous argument 'origin..HEAD': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
```

File: test/lint/lint-git-commit-check.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+echo "No commit message formatting errors detected."
```
Comment (adamjonas) [inline test/lint/lint-git-commit-check.sh:27]: If successfully passed, you'd want this to exit silently. Following the format of the other linters, I'd suggest replacing this line:
```suggestion
exit ${EXIT_CODE}
```

**Review by adamjonas — COMMENTED** [review]
Warm welcome @Ghorbanian and excited to see more contributions for you in the future!

File: test/lint/lint-git-commit-check.sh [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+# Linter to check that commit messages have a new line before the body
+# or no body at all
+#
+# reference: gist.github.com/agnivade/67b42d664ece2d4210c7
+#
+# WARNING: Using a "+" character in commit messages can result in 
+# improper parsing
+
+export LC_ALL=C
+while read -d "+" -r msg  || [[ -n "$msg" ]]; do
```
Comment (troygiorshev) [inline test/lint/lint-git-commit-check.sh:15]: This "+" idea is clever!  However, I don't think it's one we can live with.  The "+" character is actually used surprisingly often in the bodies of commit messages.

`git log --grep='+' --oneline | wc -l`
\> 706

Two suggested solutions, if you're looking for somewhere to start
- Try and find a list of valid characters in commit messages, and see if you can use one that isn't in the list.  Maybe "#" would work?  I know that emojis work in commit messages, so maybe this isn't a viable option, but check it out :)
 - Try iterate over each commit, as opposed to getting them all at once.

File: test/lint/lint-git-commit-check.sh (hunk 1 above)
Comment (troygiorshev) [inline test/lint/lint-git-commit-check.sh:25]: Check out what this outputs, I don't think it's what you want.
Possibly you're looking for `--format=%B+`, and then see if you can think of a way to deal with the "+" issue.

File: test/lint/lint-git-commit-check.sh [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    n_line=0
+    while IFS= read -r line || [[ -n "$line" ]]; do
+        n_line=$((n_line+1))
+        length=${#line}
+        if [ $n_line -eq 2 ] && [ $length -ne 0 ]; then
+            echo "Your subject line follows a non-empty line. Subject lines should always be followed by a blank line."
```
Comment (troygiorshev) [inline test/lint/lint-git-commit-check.sh:21]: This message isn't quite right.  Maybe something like one of these is better.

```suggestion
            echo "Your body follows a non-empty line. Subject lines should always be followed by a blank line."
```

```suggestion
            echo "Your subject line is followed by a non-empty line. Subject lines should always be followed by a blank line."
```

File: test/lint/lint-git-commit-check.sh [hunk 5]
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env bash
+# Copyright (c) 2017-2019 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+#
+# Linter to check that commit messages have a new line before the body
+# or no body at all
+#
+# reference: gist.github.com/agnivade/67b42d664ece2d4210c7
```
Comment (troygiorshev) [inline test/lint/lint-git-commit-check.sh:9]: Thanks for including this reference!  Let's keep it out of the code though, can you move it to the body of your commit message?

**Review by troygiorshev — COMMENTED** [review]
Hi @Ghorbanian thanks for picking this up!

When manually testing this script, I don't think it actually works :/

I've suggested below why I think this may be happening.

File: test/lint/lint-git-commit-check.sh [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+#
+# WARNING: Using a "+" character in commit messages can result in
```
[the remark this answers is not in the document]
Comment (Ghorbanian) [reply test/lint/lint-git-commit-check.sh]: Fixed.

File: test/lint/lint-git-commit-check.sh [hunk 7]
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env bash
+# Copyright (c) 2017-2019 The Bitcoin Core developers
```
[the remark this answers is not in the document]
Comment (Ghorbanian) [reply test/lint/lint-git-commit-check.sh]: fixed

File: test/lint/lint-git-commit-check.sh (hunk 1 above)
Comment (Ghorbanian) [reply test/lint/lint-git-commit-check.sh]: I changed it to `git log origin/master..HEAD`. My intent is to get the last N commits that have not yet been pushed in a portable way. I did some research and I believe the change that I have made achieves this, but please let me know if not.

File: test/lint/lint-git-commit-check.sh (hunk 2 above)
Comment (Ghorbanian) [reply test/lint/lint-git-commit-check.sh]: fixed

File: test/lint/lint-git-commit-check.sh (hunk 3 above)
Comment (Ghorbanian) [reply test/lint/lint-git-commit-check.sh]: After looking at the script some more, I decided to iterate over each commit by hash value in order to circumvent this delimiter problem. Thank you for the suggestion.

File: test/lint/lint-git-commit-check.sh (hunk 1 above)
Comment (Ghorbanian) [reply test/lint/lint-git-commit-check.sh]: fixed

File: test/lint/lint-git-commit-check.sh (hunk 4 above)
Comment (Ghorbanian) [reply test/lint/lint-git-commit-check.sh]: fixed

File: test/lint/lint-git-commit-check.sh (hunk 5 above)
Comment (Ghorbanian) [reply test/lint/lint-git-commit-check.sh]: fixed

**Review by am1r021 — COMMENTED** [review]
@troygiorshev Thank you for the code review!

File: test/lint/lint-git-commit-check.sh [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    while IFS= read -r line || [[ -n "$line" ]]; do
+        n_line=$((n_line+1))
+        length=${#line}
+        if [ $n_line -eq 2 ] && [ $length -ne 0 ]; then
+            echo "Your body is followed by a non-empty line. Subject lines should always be followed by a blank line."
+            EXIT_CODE=1
+            break
+        fi
+    done < <(git log --format=%B -n 1 "$commit_hash")
+done < <(git log origin/master..HEAD --format=%H)
```
Comment (MarcoFalke) [inline test/lint/lint-git-commit-check.sh:24]: Could make sense to default this to `TRAVIS_COMMIT_RANGE` if available?

Comment (Ghorbanian) [reply test/lint/lint-git-commit-check.sh]: fixed

**Review by troygiorshev — CHANGES_REQUESTED** [review]
Below \\/

File: test/lint/lint-git-commit-check.sh [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  fi
+fi
+
+while IFS= read -r commit_hash  || [[ -n "$commit_hash" ]]; do
+    n_line=0
+    while IFS= read -r line || [[ -n "$line" ]]; do
+        n_line=$((n_line+1))
+        length=${#line}
+        if [ $n_line -eq 2 ] && [ $length -ne 0 ]; then
+            echo "The body of commit hash ${commit_hash} is followed by a non-empty line. Subject lines should always be followed by a blank line."
```
Comment (troygiorshev) [inline test/lint/lint-git-commit-check.sh:40]: Good idea adding the commit hash!  The grammar still isn't quite what I'd like.

```suggestion
            echo "The subject line of commit hash ${commit_hash} is followed by a non-empty line. Subject lines should always be followed by a blank line."
```
or, if you like it better
```suggestion
            echo "The subject line of commit hash ${commit_hash} is followed by a non-empty line. Please add a blank line between the subject line and the body of the commit message."
```

**Review by troygiorshev — CHANGES_REQUESTED** [review]
Reviewed, manually tested. Works great!

Just need the message changed.

**Comment by troygiorshev** [comment]
ACK 284a969cc082ae3c63ab523f22e71da86ad4ab20

**Comment by fjahr** [comment]
tested ACK 1b6e02448bcced186160815a986d65ca38144940

<details><summary>Test</summary>
<p>

```
$ touch foo
$ git add foo
$ git commit -am "Foo
→ Bar"
$ test/lint/lint-git-commit-check.sh
The subject line of commit hash 1b6e02448bcced186160815a986d65ca38144940 is followed by a non-empty line. Subject lines should always be followed by a blank line.
$ test/lint/lint-git-commit-check.sh 0
$ test/lint/lint-git-commit-check.sh 1
The subject line of commit hash 1b6e02448bcced186160815a986d65ca38144940 is followed by a non-empty line. Subject lines should always be followed by a blank line.
```

</p>
</details>

**Comment by adamjonas** [comment]
utACK 284a969cc082ae3c63ab523f22e71da86ad4ab20

**Comment by fjahr** [comment]
The linter seems to fail in the CI for commit messages without a body, at least it did for me here https://travis-ci.org/github/bitcoin/bitcoin/builds/714296380 and I can not reproduce it locally. Can somebody else confirm this or is able to reproduce this locally?

**Comment by fjahr** [comment]
> The linter seems to fail in the CI for commit messages without a body, at least it did for me here https://travis-ci.org/github/bitcoin/bitcoin/builds/714296380 and I can not reproduce it locally. Can somebody else confirm this or is able to reproduce this locally?

My initial analysis of what caused the failure was wrong so ignore that part. But I think I have found what the issue was and proposed a fix here: https://github.com/bitcoin/bitcoin/pull/19654

**Pull request merged**